feat: Sequential beam search(a.k.a Low-memory beam search)

Saibo-creator opened the pull request:
# What does this PR do?
This is to address issue in https://github.com/huggingface/transformers/issues/22639 
This PR is based on the idea from
https://github.com/huggingface/transformers/issues/22639#issuecomment-1507525155

The original implementation of beam search effectively multiplies the batch size memory-wise and compute-wise by the batch size. If you have a batch size of 1 and a beam search of 8, model.forward sees 8 samples as effective batch size. This implementation is not necessary per se and can consume a lot of memory.

The new implementation split the full_batch(num_beam x batch size) inputs into a list of reduced_batch(beam_search_batch_size), run them sequentially and concat them back to a single model_output object. It involves two helper function:
-  `def concat_model_outputs(objs: List[ModelOutput]) -> ModelOutput` 
-  `def split_model_inputs( obj: Union[ModelOutput, Dict], split_size: int, full_batch_size: int ) -> List[Union[ModelOutput, Dict]]`

The new implementation can be used in 4 decoding methods: `beam_search`, `beam_sample`, `group_beam_search` and `constrained_beam_search`

The expected behavior is that it produces exactly the same output(logits) as the original implementation.

I tested it with the following quick test and it works regardless of input

```python

from
transformers import GPT2Tokenizer, AutoModelForCausalLM
import numpy as
np
from transformers import (
    AutoTokenizer,

AutoModelForSeq2SeqLM,
    LogitsProcessorList,

MinLengthLogitsProcessor,
    BeamSearchScorer,
)

tokenizer =
GPT2Tokenizer.from_pretrained("gpt2")
model =
AutoModelForCausalLM.from_pretrained("gpt2")
tokenizer.pad_token_id =
tokenizer.eos_token_id
model_inputs = tokenizer('I enjoy walking with my
cute dog', return_tensors='pt')

# activate beam search and
early_stopping
beam_output = model.generate(
    **model_inputs,

max_new_tokens=40,
    num_beams=5,

early_stopping=True
)

print("Output:\n" + 100 *
'-')
print(tokenizer.decode(beam_output[0],
skip_special_tokens=True))

beam_output_w_subbatch = model.generate(

**model_inputs,
    max_new_tokens=40,
    num_beams=5,

early_stopping=True,
    beam_search_batch_size=1
)

print("Output:\n" +
100 * '-')
print(tokenizer.decode(beam_output_w_subbatch[0],
skip_special_tokens=True))

assert (beam_output ==
beam_output_w_subbatch).all(), "Beam search results from sub batch and
full batch are different"
```

TODO: only did it for pytorch models, if you think this PR is promising, I can do it for tensorflow too.

# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->


Fixes #22639 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@gante @ArthurZucker

gante reviewed (commented):
Hey @Saibo-creator 👋 My apologies for the delayed response, I was out of office 🌴 

Thank you for opening this PR! Looking at the code, in principle, it looks good to me. The only thing I would change is the flag to trigger this low-memory behavior. We have a flag for that in contrastive search, so we could [repurpose](https://huggingface.co/docs/transformers/main_classes/text_generation#transformers.GenerationConfig.low_memory) it -- then it is `True`, we would run one beam at a time. That way we avoid adding a new argument to `generate` (which already has too many arguments).

If you update your code to reuse this existing flag, I'd be happy to approve the PR :) I suspect the code could also be simplified as a result of this change.

P.S.: a test is also missing, to confirm that low-memory code path yields the same results. LMK if you need a hand with that :)

Saibo-creator commented:
> repurpose

Hey @gante , sorry I just saw your message. I agree with your suggestion and I will update the code and add a test!

Saibo-creator commented:
Hello @gante ,
I have reused the flag `low_memory` in the code and try to unify part of the code used in sequential beam search and sequential contrastive search. 
A test is also added.

What makes me confused is the test case where I failed:
```terminal
self = <tests.models.xglm.test_modeling_xglm.XGLMModelTest testMethod=test_tf_from_pt_safetensors>

    @is_pt_tf_cross_test
    def test_tf_from_pt_safetensors(self):
        for model_class in self.all_model_classes:
            config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
    
            tf_model_class_name = "TF" + model_class.__name__  # Add the "TF" at the beginning
            if not hasattr(transformers, tf_model_class_name):
                # transformers does not have this model in TF version yet
                return
    
            tf_model_class = getattr(transformers, tf_model_class_name)
    
            pt_model = model_class(config)
    
            with tempfile.TemporaryDirectory() as tmpdirname:
                pt_model.save_pretrained(tmpdirname, safe_serialization=True)
                tf_model_1 = tf_model_class.from_pretrained(tmpdirname, from_pt=True)
    
                pt_model.save_pretrained(tmpdirname, safe_serialization=False)
                tf_model_2 = tf_model_class.from_pretrained(tmpdirname, from_pt=True)
    
                # Check models are equal
                for p1, p2 in zip(tf_model_1.weights, tf_model_2.weights):
>                   self.assertTrue(np.allclose(p1.numpy(), p2.numpy()))
E                   AssertionError: False is not true

tests/test_modeling_common.py:3246: AssertionError


FAILED tests/models/speech_to_text/test_modeling_speech_to_text.py::Speech2TextModelTest::test_tf_from_pt_safetensors - AssertionError: False is not true
FAILED tests/models/transfo_xl/test_modeling_transfo_xl.py::TransfoXLModelTest::test_tf_from_pt_safetensors - AssertionError: False is not true
FAILED tests/models/xglm/test_modeling_xglm.py::XGLMModelTest::test_tf_from_pt_safetensors - AssertionError: False is not true

Exited with code exit status 255
```

This doesn't seem to be directly related to my changes. Maybe you know what's wrong here ;D

gante commented on src/transformers/generation/utils.py:
```diff
@@ -1839,6 +1839,7 @@ def generate(
                 output_scores=generation_config.output_scores,
                 return_dict_in_generate=generation_config.return_dict_in_generate,
                 synced_gpus=synced_gpus,
+                sequential=generation_config.low_memory,
```
We should update the documentation regarding `generation_config.low_memory`, to mention that it is also compatible with beam methods :) (see [here](https://github.com/huggingface/transformers/blob/a85ea4b19a47824a8e85d3304a698e2d5c8325ec/src/transformers/generation/configuration_utils.py#L195))

gante commented on tests/generation/test_utils.py:
```diff
@@ -2637,6 +2637,21 @@ def test_beam_search_example_integration(self):
 
         self.assertListEqual(outputs, ["Wie alt bist du?"])
 
+    def test_beam_search_low_memory(self):
```
Instead of adding an integration test, I'd like to request adding a mixin -- like [this one](https://github.com/huggingface/transformers/blob/a85ea4b19a47824a8e85d3304a698e2d5c8325ec/tests/generation/test_utils.py#L1470) for contrastive search with low memory.

That way, we can ensure it is fully compatible with all models :) Like in the test linked above, it is okay to skip some models if it is failing.

Saibo-creator replied:
Sure, sounds good, I will do it ;)

One question, should I add `@slow ` for this mixin test ?

gante replied:
@Saibo-creator no need to add the `@slow` decorator -- we reserve that decorator to tests that use large models. In the mixing tests we are using small and randomly initialized models :)

gante commented on src/transformers/generation/utils.py:
```diff
@@ -4923,3 +4994,106 @@ def _ranking_fast(
+    """
+    # Infer the class from the object
+    cls = type(obj)
+    assert (full_batch_size % split_size) == 0, "`full_batch_size` must be divisible by `split_size`"
```
We avoid using `assert` in `transformers`, raising the appropriate exception instead (e.g. `ValueError`)

gante commented on src/transformers/generation/utils.py:
```diff
@@ -4923,3 +4994,106 @@ def _ranking_fast(
+    # Helper function to split tensors or tuples of tensors
+    def _split(attr):
+        """
+        Takes care three cases:
```
```suggestion
        Takes care of three cases:
```

gante reviewed (commented):
Happy to approve the PR as soon as we replace the added test by a mixin test :)

Saibo-creator commented:
Hello @gante , I would like to have your opinions over these failure cases

```terminal
FAILED tests/models/blenderbot/test_modeling_blenderbot.py::BlenderbotModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[1, 51, 51, 51, 51], [1, 37, 37, 43, 43]] != [[1, 37, 37, 43, 43], [1, 74, 74, 74, 74]]
FAILED tests/models/bloom/test_modeling_bloom.py::BloomModelTest::test_beam_search_low_memory - RuntimeError: Sizes of tensors must match except in dimension 2. Expected size 2 but got size 8 for tensor number 1 in the list.
FAILED tests/models/codegen/test_modeling_codegen.py::CodeGenModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[135, 102, 167, 94, 172, 103, 209], [233, 32, 101, 213, 234, 172, 103]] != [[135, 102, 167, 94, 172, 103, 209], [233, 32, 101, 203, 147, 47, 164]]
FAILED tests/models/clvp/test_modeling_clvp.py::ClvpDecoderTest::test_beam_search_low_memory - AssertionError: Lists differ: [[38, 235, 110, 144, 159], [93, 177, 267, 144, 55]] != [[38, 235, 110, 177, 144], [93, 177, 67, 235, 116]]
FAILED tests/models/git/test_modeling_git.py::GitModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[56, 96, 46, 38, 92, 42, 1], [25, 63, 12, 22, 49, 91, 19]] != [[56, 96, 46, 38, 92, 42, 1], [25, 63, 12, 45, 72, 77, 69]]
FAILED
FAILED tests/models/gpt_neox/test_modeling_gpt_neox.py::GPTNeoXModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[6, 34, 87, 20, 93, 20, 93], [82, 10, 76, 68, 62, 48, 10]] != [[6, 34, 87, 20, 93, 20, 93], [82, 10, 76, 68, 62, 68, 62]]
FAILED tests/models/gptj/test_modeling_gptj.py::GPTJModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[84, 2, 29, 55, 59, 47, 88], [50, 23, 76, 50, 76, 16, 20]] != [[84, 2, 29, 94, 76, 16, 20], [50, 23, 76, 50, 94, 76, 50]]
FAILED
FAILED tests/models/imagegpt/test_modeling_imagegpt.py::ImageGPTModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[57, 97, 6, 36, 77, 62, 18], [11, 23, 94, 85, 88, 81, 61]] != [[57, 97, 6, 36, 77, 62, 18], [11, 23, 94, 54, 9, 72, 18]]
FAILED tests/models/llama/test_modeling_llama.py::LlamaModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[83, 48, 56, 54, 56, 54, 56], [91, 84, 33, 72, 56, 54, 56]] != [[83, 48, 56, 54, 56, 54, 56], [91, 84, 33, 72, 56, 70, 78]]
FAILED tests/models/mega/test_modeling_mega.py::MegaModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[53, 16, 33, 21, 21, 34, 70], [82, 45, 21, 21, 34, 34, 70]] != [[53, 16, 33, 21, 34, 34, 70], [82, 45, 21, 21, 34, 34, 70]]
FAILED tests/models/mistral/test_modeling_mistral.py::MistralModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[52, 24, 42, 54, 4, 34, 21], [31, 71, 18, 5, 27, 28, 52]] != [[52, 24, 42, 54, 4, 34, 21], [31, 71, 18, 5, 73, 73, 73]]
FAILED tests/models/mpt/test_modeling_mpt.py::MptModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[8, 13, 85, 19, 19, 19, 19], [35, 61, 88, 19, 19, 19, 19]] != [[8, 13, 85, 19, 19, 19, 19], [35, 61, 88, 92, 92, 92, 92]]
FAILED tests/models/persimmon/test_modeling_persimmon.py::PersimmonModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[81, 83, 11, 58, 81, 28, 26], [54, 81, 18, 38, 18, 38, 18]] != [[81, 83, 11, 85, 58, 13, 38], [54, 81, 18, 38, 18, 38, 18]]
FAILED tests/models/phi/test_modeling_phi.py::PhiModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[28, 14, 49, 19, 56, 62, 50], [9, 30, 60, 19, 19, 19, 57]] != [[28, 14, 49, 37, 88, 35, 80], [9, 30, 60, 19, 57, 9, 57]]
```

The mixin tests reveal that a subset of models will generate different outputs over `sequential beam search` and `beam search`. 

It seems all these models are recent llm, such as `phi`, `llama`, `mistral` etc. While for other models, the outputs are identical.

I have tried debugging locally, and I confirm that:
- the beam search is deterministic across multi-runs
- the sequential beam search is also deterministic across multi-runs

So indeed these models's outputs differ in the SBS vs BS

Do you have any insights into what may have caused this difference?

Thank you!

gante commented:
Hey @Saibo-creator!

Regarding the model-specific implementations, feel free to ignore them for now (including skipping the tests, as you pasted at the end of your comment) :)

Regarding flakiness: you mentioned "the tests pass 7 out of 100 times and fails the other 93 times". Is this per model, or when you test all models at once? 
👉 If it is the latter, for all models at once, then it means the per model failure rate is low -- we can simply add the `@is_flaky` test decorator, adding a comment pointing to the comment I linked. 
👉 If it is the former, per model, then the failure rate is super high! There may be something else that we must uncover before merging :)

Saibo-creator commented:
Hey @gante 

I think it's the former case, per model... 

I may need to dive into the problem. 

If you are interested,  here is what is happening, If I run `pytest --flake-finder --flake-runs=100 tests/models/llama/test_modeling_llama.py`,

 I get 

` 93 failed, 8307 passed, 4700 skipped, 1613 warnings in 930.30s (0:15:30) ` 

and the failing cases are from `test_beam_search_low_memory` like:
```
FAILED tests/models/llama/test_modeling_llama.py::LlamaModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[7, 68, 64, 28, 26, 37, 49], [52, 58, 92, 50, 7, 37, 49]] != [[7, 68, 64, 7, 37, 49, 37], [52, 58, 92, 71, 41, 7, 41]]
FAILED tests/models/llama/test_modeling_llama.py::LlamaModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[32, 79, 18, 50, 94, 48, 93], [52, 51, 3, 56, 24, 12, 24]] != [[32, 79, 18, 50, 94, 5, 25], [52, 51, 3, 56, 24, 12, 24]]
FAILED tests/models/llama/test_modeling_llama.py::LlamaModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[39, 97, 28, 57, 0, 0, 0], [34, 69, 91, 34, 63, 0, 0]] != [[39, 97, 28, 57, 0, 0, 0], [34, 69, 91, 5, 58, 18, 14]]
```

So this is 93/ 100 where the two outputs are different


For the other models such as `GPT2`, I also get som failure cases, though not frequent
with `pytest --flake-finder --flake-runs=10 tests/models/gpt2/test_modeling_gpt2.py`

FAILED tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[29, 58, 3, 68, 68, 68, 68], [90, 65, 45, 45, 68, 68, 68]] != [[29, 58, 3, 68, 68, 68, 68], [90,...
FAILED tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[18, 32, 29, 29, 87, 87, 87], [40, 85, 64, 64, 30, 30, 30]] != [[18, 32, 29, 87, 87, 87, 87], [4...
FAILED tests/models/gpt2/test_modeling_gpt2.py::GPT2ModelTest::test_beam_search_low_memory - AssertionError: Lists differ: [[98, 79, 92, 9, 9, 9, 9], [46, 16, 93, 9, 9, 9, 9]] != [[98, 79, 92, 9, 9, 9, 9], [46, 16, 93, 9...
=============================== 3 failed, 907 passed, 510 skipped, 197 warnings in 470.60s (0:07:50)

gante commented:
@Saibo-creator sadly I have no easy advice here :( it could be a kernel-related numerical issue due to different shapes, like when using `past_key_values`, or it can be a subtle bug.

I see to paths forward:
1 - you are able to pin the source of the mismatch, and we can safely confirm that it is something like a numerical issue
2 - you run a few benchmarks over your PR -- if the resulting metrics are similar with the low memory mode, then there shouldn't be a bug

Saibo-creator commented:
I spent some time investigating the reason for mismatch. I notice that if batch-size = 1, all tests pass without any failure(as shown in the latest commit)

So it seems this is not a numerical issue but more like something to do with batch processing. 

I will try to figure out if my implementation has a bug with batched input.

Saibo-creator commented:
I confirm that if we disable `use_cache`, the output are identical. 
So now we identified the origin was from the `key_value_cache`

izyForever committed:
update the logger message with accordant weights_file_name (#28181)

Co-authored-by: yudong.lin <yudong.lin@funplus.com>

Saibo-creator committed:
add sequential beam search (clean version)

Saibo-creator committed:
fix bug in key_value split, this solves the problem of output mismatch

Saibo-creator commented:
Hello @gante ，
     After a whole afternoon debugging, I found the bug, it's a tiny bug [in this line](https://github.com/epfl-dlab/transformers-GCD-PR/blob/fix_issue_22639/src/transformers/generation/utils.py#L4955). I somehow made the `for i in range(0, full_batch_size, 1)` instead of `for i in range(0, full_batch_size, split_size)`. The former makes no sense at all so I guess it was just a typo, but hidden very well....

    I'm happy that it works now : )

younesbelkada committed:
[`Llava`] Fix llava index errors (#28032)

* fix llava index errors

* forward contrib credits from original implementation and fix

* better fix

* final fixes and fix all tests

* fix

* fix nit

* fix tests

* add regression tests

---------

Co-authored-by: gullalc <gullalc@users.noreply.github.com>

younesbelkada committed:
[`Awq`] Enable the possibility to skip quantization for some target modules (#27950)

* v1

* add docstring

* add tests

* add awq 0.1.8

* oops

* fix test

pacman100 committed:
fix FA2 when using quantization (#28203)

stas00 committed:
small typo (#28229)

Update modeling_utils.py

dwyatte committed:
Update docs around mixing hf scheduler with deepspeed optimizer (#28223)

update docs around mixing hf scheduler with deepspeed optimizer

hiyouga committed:
Fix trainer saving safetensors: metadata is None (#28219)

* Update trainer.py

* format

frankenliu committed:
fix bug:divide by zero in _maybe_log_save_evaluate() (#28251)

Co-authored-by: liujizhong1 <liujizhong1@xiaomi.com>

ercaronte committed:
[Whisper] Fix errors with MPS backend introduced by new code on word-level timestamps computation (#28288)

* Update modeling_whisper.py to support MPS backend

Fixed some issue with MPS backend.

First, the torch.std_mean is not implemented and is not scheduled for implementation, while the single torch.std and torch.mean are.
Second, MPS backend does not support float64, so it can not cast from float32 to float64. Inverting the double() when the matrix is in the cpu fixes the issue while should not change the logic.

* Found another instruction in modeling_whisper.py not implemented byor MPS

After a load test, where I transcribed a 2 hours audio file, I got into a branch that did not fix in the previous commit.
Similar fix, where the torch.std_mean is changed into torch.std and torch.mean

* Update modeling_whisper.py removed trailing white spaces

Removed trailing white spaces

* Update modeling_whisper.py to use is_torch_mps_available()

Using is_torch_mps_available() instead of capturing the NotImplemented exception

* Update modeling_whisper.py sorting the import block

Sorting the utils import block

* Update src/transformers/models/whisper/modeling_whisper.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

* Update src/transformers/models/whisper/modeling_whisper.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

* Update src/transformers/models/whisper/modeling_whisper.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

---------

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

dbuos committed:
Remove fast tokenization warning in Data Collators (#28213)

dependabot[bot] committed:
Bump tj-actions/changed-files from 22.2 to 41 in /.github/workflows (#28311)

Bumps [tj-actions/changed-files](https://github.com/tj-actions/changed-files) from 22.2 to 41.
- [Release notes](https://github.com/tj-actions/changed-files/releases)
- [Changelog](https://github.com/tj-actions/changed-files/blob/main/HISTORY.md)
- [Commits](https://github.com/tj-actions/changed-files/compare/v22.2...v41)

---
updated-dependencies:
- dependency-name: tj-actions/changed-files
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

gante committed:
Merge branch 'main' into fix_issue_22639

gante reviewed (approved):
LGTM, thank you for iterating (and for nailing the issue in the previous commit!) 💛

ArthurZucker commented on src/transformers/generation/utils.py:
```diff
@@ -4905,3 +4914,124 @@ def _ranking_fast(
+                    tuple(sub_tensor[i : i + split_size] for sub_tensor in data)
+                    for i in range(0, full_batch_size, split_size)
+                ]
+        elif isinstance(data, ModelOutput):
+            return _split_model_inputs(model_output["encoder_outputs"], split_size, full_batch_size)
```
bit strange to define and call in the same function let's maybe define `_split` outside

ArthurZucker replied:
I am also not a super fan of the nested model outputs being handled that way, we should IMO have something like 
```python 
def seperate_model_outputs(model_outputs):
      if "encoder_outputs" in model_outputs:
             model_outputs["encoder_outputs"] = split(model_outputs["encoder_outputs"])
      model_outputs = split(model_outputs) # ignores the ModelOutput type that is already split
      return model_outputs
```
since the only model output you ever have is the `encoder_outputs` this to me makes it more readable and clearer! 
It adds an extra function but should make it easier to maintain 🤗

ArthurZucker commented on src/transformers/generation/utils.py:
```diff
@@ -4905,3 +4914,124 @@ def _ranking_fast(
+    # bool should not be split but replicated for each split
+    bool_keys = [k for k in keys if isinstance(model_output[k], bool)]
+    non_bool_keys = [k for k in keys if not isinstance(model_output[k], bool)]
+    # import pdb; pdb.set_trace()
```
```suggestion
```

ArthurZucker reviewed (commented):
Thanks for this huge feature! 
 #27539 added some support for 4D attention masks which support packing, not the scope but a bit related! 
LGTM other than a small nit 🤗

JulesGM commented:
@Saibo-creator any reason this hasn't been merged?

JulesGM commented:
or @gante ?

JulesGM commented:
Thanks for implementing it btw @Saibo-creator

JulesGM commented:
I could really use this

Saibo-creator commented:
> @Saibo-creator any reason this hasn't been merged?

Hello Jules! Sorry I was a bit out of bandwidth last week, I'm fixing the small things pointed by Arthur now and will push immediately : )

Saibo-creator committed:
refactor _split_model_inputs function for sequential beam search.

Saibo-creator committed:
fix error with test_modeling_whisper.py

Reason is because encoder_outputs from whisper is None in this test

Saibo-creator committed:
fix syntax error(feature only available in python3.10)

Saibo-creator committed:
fix typo

Saibo-creator commented:
@ArthurZucker  issues solved. Let me know any other things to do

JulesGM commented:
The code automatically taking the batch size of the input_ids makes it pretty awkward to modify existing code, as it changes the concept of what a batch is compared to other generation types, and requires some weird batch break-down and recomposition in eval code. If you don't want to add a flag, maybe batch_size could be passed as the value of `low_memory` ?

JulesGM commented:
As it is, it is still a great improvement, though.

JulesGM commented:
Is there a way the batch size of the beam search could be unbounded from the `num_return_sequences`?

JulesGM commented:
For example, for RL, I would like to generate let's say 64 different outputs with a single input sequence. However, my hardware (multiple gpu or not) hardware can't handle more than 16 beams at once.

JulesGM commented:
From trying to run the code, it seems like I couldn't get those 64 outputs? Just 16?

ArthurZucker reviewed (approved):
LGTM, thanks for you patience!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26304). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

gante commented:
Merging as it's approved as it solves the original target use case.

@JulesGM don't refrain from commenting if it doesn't solve your use case: we may still open a follow-up PR to further refine this 🤗

gante commented:
@Saibo-creator thank you for iterating with us 💛

Saibo-creator commented:
> Merging as it's approved as it solves the original target use case.
> 
> @JulesGM don't refrain from commenting if it doesn't solve your use case: we may still open a follow-up PR to further refine this 🤗

@JulesGM If you think it's necessary,  we could consider adding more flexibility by allowing users to specify the `parallel_size` for example. It would be an easy add-on.

gante commented:
@Saibo-creator alternatively, a try/except loop could be tried (try with `batch_size` -> if it fails, try with `batch_size/2` -> ... -> try with batch_size = 1). I'd prefer that to an additional flag -- we already have too many flags in `generate` :)

Saibo-creator commented:
> @Saibo-creator alternatively, a try/except loop could be tried (try with `batch_size` -> if it fails, try with `batch_size/2` -> ... -> try with batch_size = 1). I'd prefer that to an additional flag -- we already have too many flags in `generate` :)

Thanks for the suggestion @gante  ! I think this would be a good improvement to reduce work from user side. Should I open another PR to proceed ?

JulesGM commented:
> > Merging as it's approved as it solves the original target use case.
> > @JulesGM don't refrain from commenting if it doesn't solve your use case: we may still open a follow-up PR to further refine this 🤗
> 
> @JulesGM If you think it's necessary, we could consider adding more flexibility by allowing users to specify the `parallel_size` for example. It would be an easy add-on.

I think that this would be great. In my personal fork, I currently look at if low_memory is an int, and use it as a sub-batch size if it is the case ahah.

JulesGM commented:
I also extended the code in my personal fork to beam_sample, basically copy-pasting the beam_search sequential section. I think that that would be a good thing to add.

gante commented:
@Saibo-creator yeah, open a new PR, I'd be happy to merge it 🤗

Saibo-creator commented:
> @Saibo-creator yeah, open a new PR, I'd be happy to merge it 🤗 

👍🏻 I will make a PR shortly

JulesGM commented:
So I noticed that in practice this didn't allow me to increase the batch size by that much if the sequence size is long. So I'm experimenting with offloading the key value cache to cpu & only loading the part of the cache that is needed, & it seems to really help.

Saibo-creator commented:
> So I noticed that in practice this didn't allow me to increase the beam size by that much at all if the sequence size is long. So I'm experimenting with offloading the key value cache to cpu & only loading the part of the cache that is needed, & it seems to really help.

Oh, thanks for pointing out @JulesGM . I think your observation makes sense. Offloading the kv-cache should indeed be more efficient!

Saibo-creator commented:
> So I noticed that in practice this didn't allow me to increase the beam size by that much at all if the sequence size is long. So I'm experimenting with offloading the key value cache to cpu & only loading the part of the cache that is needed, & it seems to really help.

Hey @JulesGM , this approach is cool,  would you like to share a link to your implementation of this feature? I could benefit from it too :)

thanhlecongg commented:
Hi @Saibo-creator, I have tried to run the model with `low_memory=True` with high. However, it is quite counter-intuitive that I observed memory usage than the original one. 

`
### Low_memory = True
Memory usage:  19.74514389038086
Time taken:  63.430684089660645


### Normal
Memory usage:  10.743547439575195
Time taken:  2.9829838275909424
`

This is my code:

`
import torch
from transformers import GPT2Tokenizer, AutoModelForCausalLM
import numpy as np
import time
from transformers import (
    AutoTokenizer,

AutoModelForSeq2SeqLM,
    LogitsProcessorList,

MinLengthLogitsProcessor,
    BeamSearchScorer,
)

def get_memory_usage():
    return torch.cuda.max_memory_allocated(device=None)/1024/1024/1024

tokenizer = GPT2Tokenizer.from_pretrained("gpt2")
model = AutoModelForCausalLM.from_pretrained("gpt2")
model.to('cuda')
tokenizer.pad_token_id = tokenizer.eos_token_id
model_inputs = tokenizer('I enjoy walking with my cute dog' * 50, return_tensors='pt')
model_inputs.to('cuda')

start = time.time()

beam_output = model.generate(**model_inputs,
                            max_new_tokens=40,
                            num_beams=100,
                            early_stopping=True
)

print("Output:\n" + 100 *'-')
print(tokenizer.decode(beam_output[0], skip_special_tokens=True))

# Uncomment for running low_memory=True
# beam_output_w_subbatch = model.generate(
#     **model_inputs,
#     max_new_tokens=40,
#     num_beams = 100,
#     early_stopping = True,
#     low_memory = True
# )

# print("Output:\n" + 100 * '-')
# print(tokenizer.decode(beam_output_w_subbatch[0], skip_special_tokens=True))

print("Memory usage: ", get_memory_usage())
print("Time taken: ", time.time() - start)
`

I'm thinking that it is because you are stacking outputs on these [lines](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L3231-L3235) before releasing memory. So it consume double memory (both in outputs_per_sub_batch and outputs) than expected. Do you have any thoughts on this?

Thanks.

Saibo-creator commented:
Hello @thanhlecongg , Thanks for the question!

You are looking at the right [place](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/utils.py#L3231-L3235). 


## Why is the memory usage higher in `low_memory=True` than the vanilla beam search?
The operation of stacking `outputs_per_sub_batch` effectively doubles the memory usage of the model output.

In cases where the model output is small, this is not a problem. But in cases where the model output is long ( your case), the memory usage can overwhelm the memory usage for computing the model output.

For example, with beam size = 20, sequence length = 350, the `model_output.logits` is 20 x 350 x vocab_size(50K) x 4 bytes ~ 1.4GB. 


## This is counter-intuitive, when `low_memory=True` is supposed to reduce memory usage ?
This is not a big problem per se because depending on the use case, the output may not be huge and the memory bottleneck is the model size and the intermediate variables in forward pass.

One example of a use case where `low_memory=True` is useful is when the model is very large and the input is relatively short,
for example, use
-  `gpt2-xl`(1.2 Trillion params) instead of `gpt2` (150 Million Params)
-  `I enjoy walking with my cute dog * 2` instead of `I enjoy walking with my cute dog * 50`
- beam_size=4

In this case, the vanilla beam search will OOM (my GPU is 8GB) but `low_memory=True` will work.


## Fix 
One simple fix is to do the **stacking** operation in CPU memory and then move it back to GPU memory.

This will make sure the memory usage in case of `low_memory=True` is always less than the vanilla beam search.


@gante , do you think this is a valid issue? If yes, I can create a PR for this.

thanhlecongg commented:
Thanks for your quick response and clear explaination.
